slices/filter_map.go

```diff
@@ -0,0 +1,49 @@
+package slices
+
+import "github.com/akitasoftware/go-utils/optionals"
+
+// Applies f to each element of slice in order, removes any None results, and
+// returns the rest.
+func FilterMap[T1, T2 any](slice []T1, f func(T1) optionals.Optional[T2]) []T2 {
+	result, _ := FilterMapIndexWithErr(slice, func(_ int, t T1) (optionals.Optional[T2], error) {
+		return f(t), nil
+	})
+	return result
+}
+
+// Applies f to each element of slice in order, removes any None results, and
+// returns the rest. Returns an error if f returns a non-nil error on any
+// element.
+func FilterMapWithErr[T, U any](slice []T, f func(T) (optionals.Optional[U], error)) ([]U, error) {
```

Comment:
Maybe I'm being dense, but why do the *WithErr variants use T and U instead of the T1 and T2 type variables used by FilterMap and FilterMapIndexWithErr?

Reply:
No particular reason. They happened to be written at a different time than other variants.

What's in a type parameter name? That which we call a T1 by any other name would accept just as many types.

Well, that didn't work very well. Here's ChatGPT's attempt, which seems to mix in some Hamlet:

>> Adapt Shakespeare's "What's in a name" soliloquy for type variables.
>
> Alas, poor type variable! What strife
Thou bring'st upon our coding life!
What madness doth thy name create?
What errors do we perpetrate?
>
>Forsooth, a type by any name
Would function just the same,
But in thy naming lies the rub,
And with it comes a lot of hubbub.
>
>Doth "T" or "U" or "V" declare
The nature of its form and air?
Nay, 'tis but empty syntax, void
Of any meaning to avoid.
>
>Yet still we strive to find a word
To match the concept we've inferred,
To give thee substance and a face,
And thus we fall from grace.
>
>So what's in a type variable's name?
Just endless trials and endless game.
Let us seek not for fancy phrase
But use simplicity to amaze.

Reply:
Renamed in c8cc8a8.